rocAL/rocAL/source/sequence_file_source_reader.cpp

```diff
@@ -0,0 +1,306 @@
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include <cassert>
+#include <algorithm>
+#include <commons.h>
+#include "sequence_file_source_reader.h"
+#include <boost/filesystem.hpp>
+
+namespace filesys = boost::filesystem;
+
+SequenceFileSourceReader::SequenceFileSourceReader() : _shuffle_time("shuffle_time", DBG_TIMING)
+{
+    _src_dir = nullptr;
+    _sub_dir = nullptr;
+    _entity = nullptr;
+    _curr_file_idx = 0;
+    _current_file_size = 0;
+    _current_fPtr = nullptr;
+    _loop = false;
+    _sequence_id = 0;
+    _shuffle = false;
+    _sequence_count_all_shards = 0;
+}
+
+unsigned SequenceFileSourceReader::count()
+{
+    if (_loop)
+        return _frame_names.size();
+
+    int ret = ((int)_frame_names.size() - _read_counter);
+    return ((ret < 0) ? 0 : ret);
+}
+
+Reader::Status SequenceFileSourceReader::initialize(ReaderConfig desc)
+{
+    auto ret = Reader::Status::OK;
+    _sequence_id = 0;
+    _folder_path = desc.path();
+    _shard_id = desc.get_shard_id();
+    _shard_count = desc.get_shard_count();
+    _user_batch_count = desc.get_batch_size();
+    _shuffle = desc.shuffle();
+    _loop = desc.loop();
+    _sequence_length = desc.get_sequence_length();
+    _step = desc.get_frame_step();
+    _stride = desc.get_frame_stride();
+    _batch_count = _user_batch_count / _sequence_length;
+
+    ret = subfolder_reading();
+    ret = get_sequences();
+
+    // the following code is required to make every shard the same size:: required for multi-gpu training
+    if (_shard_count > 1 && _batch_count > 1)
+    {
+        int _num_batches = _sequence_frame_names.size() / _batch_count;
+        int max_batches_per_shard = (_sequence_count_all_shards + _shard_count - 1) / _shard_count;
+        max_batches_per_shard = (max_batches_per_shard + _batch_count - 1) / _batch_count;
+        if (_num_batches < max_batches_per_shard)
+        {
+            replicate_last_batch_to_pad_partial_shard();
+        }
+    }
+    //shuffle dataset if set
+    _shuffle_time.start();
+    if (ret == Reader::Status::OK && _shuffle)
+        std::random_shuffle(_sequence_frame_names.begin(), _sequence_frame_names.end());
+    _shuffle_time.end();
+
+    for (unsigned i = 0; i < _sequence_frame_names.size(); i++)
+    {
+        for (unsigned j = 0; j < _sequence_frame_names[i].size(); j++)
+        {
+            _frame_names.push_back(_sequence_frame_names[i][j]);
```

Comment:
why there needs to be 2 different structures for storing seq names?

Reply:
_sequence_frame_names is a 2D vector , with the outermost vector representing number of sequences and each element will contain sequence length number of frame names (this is required for shuffling where we will need to shuffle the sequences and not the frames names). On the other hand _frame_names is a 1D representation of all the frame names.

Reply:
When we shuffle, isn't it supposed to shuffle the sequences? Is it required to shuffle frames within a sequence as well?

Reply:
Yes only the sequences will be shuffled (_sequence_frame_names - 2D vector of frame names of sequences)

Reply:
Looks like the sequence reader is reading each frame of the sequence one at a time. We need read an entire sequence at once since the lowest unit for sequence reader is a sequence of frames. Also storing filename for each frame of a video sequence will blow up the memory when we have to decode larger video